### PR TITLE
1559

### PR DESCRIPTION
**Pull request by MarekM25**
@tkstanczak if you want to review only the last changes you can check https://github.com/NethermindEth/nethermind/compare/xdai1559_tests...1559?expand=1

@LukaszRozmej I would like to see your review and after that, I will run some nodes with those changes schedule smoke tests, and run goerli validator

Merge plan:
- [x] Review by @LukaszRozmej 
- [x] Run some nodes
- [ ] Run smoke tests
- [x] Run Goerli Validator
- [ ] Merge

### CODE REVIEW

File: src/Nethermind/Nethermind.Runner/Ethereum/Steps/StartBlockProducerAuRa.cs [hunk 1]
```diff
@@ -98,9 +103,14 @@ protected override void BuildProducer()
...
             if (_api.EngineSigner == null) throw new StepDependencyException(nameof(_api.EngineSigner));
+            if (_api.SpecProvider == null) throw new StepDependencyException(nameof(_api.SpecProvider));
 
             var chainSpecAuRa = _api.ChainSpec.AuRa;
-            _txPermissionFilter = TxAuRaFilterBuilders.CreateTxPermissionFilter(_api, readOnlyTxProcessorSource);
+
+            ITxFilter auRaTxFilter = TxAuRaFilterBuilders.CreateAuRaTxFilter(
+                _api,
+                readOnlyTxProcessorSource,
+                _api.SpecProvider);;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Runner/Ethereum/Steps/StartBlockProducerAuRa.cs:113]: double ;

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Blockchain/Comparers/GasPriceTxComparerHelper.cs [hunk 2]
```diff
@@ -15,28 +15,35 @@
...
+using Nethermind.Int256;
 
-namespace Nethermind.TxPool
+namespace Nethermind.Blockchain.Comparers
 {
-    /// <summary>
-    /// Default ordering by <see cref="Transaction.GasPrice"/> desc
-    /// </summary>
-    public class CompareTxByGasPrice : IComparer<Transaction>
+    public static class GasPriceTxComparerHelper
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Blockchain/Comparers/GasPriceTxComparerHelper.cs:23]: Let's discuss

File: src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs [hunk 3]
```diff
@@ -160,13 +168,14 @@ bool HasEnoughFounds(IDictionary<Address, UInt256> balances, Transaction transac
                     if (_logger.IsDebug) _logger.Debug($"Rejecting (invalid nonce - expected {expectedNonce}) {tx.ToShortString()}");
                     continue;
                 }
-
-                if (!HasEnoughFounds(remainingBalance, tx))
+                
+                if (!HasEnoughFounds(remainingBalance, tx, isEip1559Enabled, _blockPreparationContextService.BaseFee))
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs:172]: Let's discuss

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Int256;
+using Nethermind.TxPool;
+
+namespace Nethermind.Consensus.AuRa.Transactions
+{
+    public class TxGasPrice1559Sender : ITxSender
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:28]: Yes. Please check TxGasPrice1559Sender references

File: src/Nethermind/Nethermind.Consensus.Ethash/NethDevPlugin.cs [hunk 5]
```diff
@@ -57,17 +58,31 @@ public Task InitBlockProducer()
...
             var (getFromApi, setInApi) = _nethermindApi!.ForProducer;
-            ITxFilter txFilter = new NullTxFilter();
-            ITxSource txSource = new TxPoolTxSource(
-                getFromApi.TxPool, getFromApi.StateReader, getFromApi.LogManager, txFilter);
 
+            ReadOnlyDbProvider readOnlyDbProvider = getFromApi.DbProvider.AsReadOnly(false);
+            ReadOnlyBlockTree readOnlyBlockTree = getFromApi.BlockTree.AsReadOnly();
+
+            ITransactionComparerProvider transactionComparerProvider =
+                new TransactionComparerProvider(getFromApi.SpecProvider, getFromApi.BlockTree);
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus.Ethash/NethDevPlugin.cs:67]: We always pass blockTree when we're expecting IBlockFinder. Changed to readonlyBlockTree. Let's discuss

File: src/Nethermind/Nethermind.Consensus/Transactions/BaseFeeTxFilter.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            _specProvider = specProvider;
+            _logger = logManager.GetClassLogger();
+        }
+
+        public (bool Allowed, string Reason) IsAllowed(Transaction tx, BlockHeader parentHeader)
+        {
+            if (parentHeader.Number + 1 != _blockPreparationContextService.BlockNumber)
+            {
+                if (_logger.IsWarn) _logger.Warn($"Wrong context was set: Context block number: {_blockPreparationContextService.BlockNumber} Parent header: {parentHeader}");
+            }
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus/Transactions/BaseFeeTxFilter.cs:46]: Let's discuss

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**